File: src/WorkBC.Notifications.JobAlerts/Services/JobAlertSearchService.cs
using System;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;
using Serilog;
using WorkBC.Data;
using WorkBC.Data.Enums;
using WorkBC.Data.Model.JobBoard;
using WorkBC.ElasticSearch.Models.Filters;
using WorkBC.ElasticSearch.Models.Results;
using WorkBC.ElasticSearch.Search.Queries;
using WorkBC.Shared.Services;

namespace WorkBC.Notifications.JobAlerts.Services
{
    /// <summary>
    ///     Wrapper around Elastic search, used by the Job Alerts scheduled task
    /// </summary>
    public class JobAlertSearchService
    {
        private readonly IConfiguration _configuration;
        private readonly IGeocodingService _geocodingService;

        public JobAlertSearchService(IConfiguration configuration, JobBoardContext dbContext)
        {
            _configuration = configuration;
            var geoService = new GeocodingService(configuration);
            _geocodingService = new GeocodingCachingService(dbContext, geoService);
        }

        /// <summary>
        ///     Searches for jobs matching the filter and posted within the job alert frequency
        /// </summary>
        public async Task<long> GetJobAlertSearchResultCount(JobAlert jobAlert)
        {
            var filters = JsonConvert.DeserializeObject<JobSearchFilters>(jobAlert.JobSearchFilters);
            filters.PageSize = 0;

            DateTime startDate = GetFilterStartDate(jobAlert.AlertFrequency);

            filters.SearchDateSelection = "3"; // custom date range
            filters.StartDate = new DateField(startDate);
            filters.EndDate = new DateField(DateTime.Now);

            //Search object that we will use to search Elastic Search
            var esq = new JobSearchQuery(_geocodingService, _configuration, filters);

            //Get search results from Elastic search
            ElasticSearchResponse results = await esq.GetSearchResults();

            long result = results?.Hits?.Total?.Value ?? 0;
            return result;
        }

        /// <summary>
        ///     Gets the start date for the filter based on frequency
        /// </summary>
        private static DateTime GetFilterStartDate(JobAlertFrequency frequency)
        {
            // set the dates
            TimeSpan removeTimeSpan;

            switch (frequency)
            {
                case JobAlertFrequency.Daily:
                    removeTimeSpan = new TimeSpan(1, 0, 0, 0);
                    break;
                case JobAlertFrequency.Weekly:
                    removeTimeSpan = new TimeSpan(7, 0, 0, 0);
                    break;
                case JobAlertFrequency.BiWeekly:
                    removeTimeSpan = new TimeSpan(15, 0, 0, 0);
                    break;
                default:
                    removeTimeSpan = new TimeSpan(31, 0, 0, 0);
                    break;
            }

            return DateTime.Now.Subtract(removeTimeSpan).Date;
        }
    }
}

File: src/WorkBC.Notifications.JobAlerts.V2/src/Service/JobAlertSearchService.php
<?php

declare(strict_types=1);

namespace WorkBC\JobAlertNotifier\Service;

use WorkBC\JobAlertNotifier\Elastic\ElasticSearchClient;
use WorkBC\JobAlertNotifier\Search\DateField;
use WorkBC\JobAlertNotifier\Search\GeocodingService;
use WorkBC\JobAlertNotifier\Search\JobSearchFilters;
use WorkBC\JobAlertNotifier\Search\JobSearchQuery;

/**
 * Port of WorkBC.Notifications.JobAlerts.Services.JobAlertSearchService —
 * the wrapper around Elasticsearch used by the job alert task.
 */
final class JobAlertSearchService
{
    private string $template;

    public function __construct(
        private readonly ElasticSearchClient $elastic,
        private readonly GeocodingService $geocodingService,
        private readonly string $index,
    ) {
        $path = dirname(__DIR__, 2) . '/resources/jobsearch_main.json.template';
        $template = file_get_contents($path);
        if ($template === false) {
            throw new \RuntimeException("Unable to read resource file: {$path}");
        }
        $this->template = $template;
    }

    /**
     * Searches for jobs matching the alert's saved filter and posted within
     * the alert frequency window; returns the total hit count.
     */
    public function getJobAlertSearchResultCount(string $jobSearchFiltersJson, int $alertFrequency): int
    {
        $filters = JobSearchFilters::fromJson($jobSearchFiltersJson);
        $filters->pageSize = 0;

        $startDate = self::getFilterStartDate($alertFrequency);

        $filters->searchDateSelection = '3'; // custom date range
        $filters->startDate = DateField::fromDateTime($startDate);
        $filters->endDate = DateField::fromDateTime(new \DateTimeImmutable('now'));

        $query = new JobSearchQuery($this->geocodingService, $filters);
        $json = $query->toJson($this->template);

        $results = $this->elastic->search($this->index, $json);

        return (int) ($results['hits']['total']['value'] ?? 0);
    }

    /**
     * Gets the start date for the filter based on frequency:
     * daily −1 day, weekly −7, biweekly −15, monthly (default) −31,
     * truncated to midnight like DateTime.Date.
     */
    private static function getFilterStartDate(int $frequency): \DateTimeImmutable
    {
        $days = match ($frequency) {
            JobAlertEmail::FREQUENCY_DAILY => 1,
            JobAlertEmail::FREQUENCY_WEEKLY => 7,
            JobAlertEmail::FREQUENCY_BIWEEKLY => 15,
            default => 31,
        };

        return (new \DateTimeImmutable("now -{$days} days"))->setTime(0, 0, 0, 0);
    }
}
